go/ql/lib/semmle/go/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -88,13 +98,24 @@ class AdditionalTaintStep extends Unit {
  */
 predicate localAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ, string model) {
   (
-    referenceStep(pred, succ) or
-    elementWriteStep(pred, succ) or
-    fieldReadStep(pred, succ) or
-    elementStep(pred, succ) or
-    tupleStep(pred, succ) or
-    stringConcatStep(pred, succ) or
+    referenceStep(pred, succ)
+    or
+    elementWriteStep(pred, succ)
+    or
+    fieldReadStep(pred, succ)
+    or
+    elementStep(pred, succ)
+    or
+    tupleStep(pred, succ)
+    or
+    stringConcatStep(pred, succ)
+    or
     sliceStep(pred, succ)
+    or
+    // Treat container flow as taint for the local taint flow relation
+    exists(DataFlow::Content c | DataFlowPrivate::containerContent(c) |
```

Comment:
What container read steps are not element steps?

Reply:
Do you mean that I should put this in `elementStep`? I am happy to do that. If you mean "what does add beyond `elementStep`" then I'd have to think about it a bit more. It certainly adds flow from an implicit varargs slice to the post-update nodes of the individual args that went into it.

Reply:
This will certainly duplicate a lot of what `elementStep` is intending to do -- but clearly "those readSteps that read array values" and "elementStep" are not equivalent, which they probably should do. I'd recommend taking a look at how/why they differ and seeing if elementStep (or perhaps sliceStep) just needs an additional clause?

Reply:
I checked and it should be `elementStep` (access an element of a container) rather than `sliceStep` (make a slice from an array/slice).
Looking carefully, `readStep` restricted to container content covers all of `elementStep`, plus the following:
1. implicit varargs slice post-update nodes, added in my recent PR (which is what we're trying to get a taint step for)
2. steps through summary models which read container content. (I think it might be okay to have taint steps for this)
3. for `a = b[i:j]`, we have a special step reading the array elements of `b`, which will match up with a corresponding special step storing those elements in `a`. (I don't think we want taint steps for this  - we already have a taint step from `b` to `a` in `sliceStep`.)
4. channel receive expressions `<- chan`. (Not sure about this.)
5. in a ranged for loop `for a, b := range c`, when `c` is a map, flow from the map to the key (with MapKey content) and to the value (with MapValue content). (We currently only have a taint step for `c` to `b`, which is what makes sense for non-maps. Not sure if adding this would be a good thing.)

Reply:
That suggests to me take care of (1) in elementStep, then (4) can be considered separately

Reply:
Done

Reply:
Yes, I'm not averse to using `readStep` to define `localAdditionalTaintStep`, but I am averse to doing so where there is a somewhat murky venn diagram of steps supplied via `readStep` and steps supplied directly by the other disjuncts in `localAdditionalTaintStep`.